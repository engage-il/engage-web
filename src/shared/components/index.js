// @flow
export { Link } from './Link'
export { CopyLink } from './CopyLink'
export { Button } from './Button'
export { NotificationView } from './NotificationView'
