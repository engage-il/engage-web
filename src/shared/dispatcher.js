// @flow
import Events from 'ampersand-events'

const dispatcher = Events.createEmitter()

type Event = 'menu-changed'
type EventHandler = (args: any) => void

export function dispatch (event: Event, args: any) {
  dispatcher.trigger(event, args)
}

export function on (event: Event, handler: EventHandler) {
  dispatcher.on(event, handler)
}

export function off (event: Event, handler: EventHandler) {
  dispatcher.off(event, handler)
}
