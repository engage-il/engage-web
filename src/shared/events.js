// @flow
import EventEmitter from 'eventemitter3'

class Events extends EventEmitter {
  setAuthHeader = '@@events/set-auth-header'
  setEnvironment = '@@events/set-environment'
  showNotification = '@@events/show-notification'
}

export const events = new Events()
