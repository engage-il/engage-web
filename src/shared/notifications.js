// @flow
export type Notification = {
  key: string,
  message: string
}

type NotificationListener = (Notification) => (void | Promise<void>)

let _listener: ?NotificationListener = null

export const notifications = {
  on (listener: NotificationListener) {
    _listener = listener
  },
  off () {
    _listener = null
  },
  add (notification: Notification) {
    _listener && _listener(notification)
  }
}
