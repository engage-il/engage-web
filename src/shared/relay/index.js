// @flow
export { currentEnvironment } from './currentEnvironment'
export { unwrap } from './unwrap'
export { withLoadMoreArgs } from './withLoadMoreArgs'
export { cacheResolvers } from './cacheResolvers'
export { createPaginationCacheResolver } from './createPaginationCacheResolver'
