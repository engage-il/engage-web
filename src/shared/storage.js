/* globals Storage, localStorage, sessionStorage */
// @flow

function namespaceKey (key: string): string {
  return `@@legislated/${key}`
}

class Store<K: string> {
  storage: Storage

  constructor (storage: Storage) {
    this.storage = storage
  }

  set (key: K, value: ?string) {
    const actualKey = namespaceKey(key)

    if (value) {
      this.storage.setItem(actualKey, value)
    } else {
      this.storage.removeItem(actualKey)
    }
  }

  get (key: K): ?string {
    const actualKey = namespaceKey(key)
    return this.storage.getItem(actualKey)
  }
}

type LocalStoreKey
  = 'intro-visited'
  | 'intro-cleared'

type SessionStoreKey
  = 'last-search-count'
  | 'admin-header'

export const local: Store<LocalStoreKey> = new Store(localStorage)
export const session: Store<SessionStoreKey> = new Store(sessionStorage)
