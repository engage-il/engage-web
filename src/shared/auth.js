// @flow
import { session } from 'shared/storage'
import { events } from 'shared/events'
import { Base64 } from 'js-base64'

export const auth = {
  get isSignedIn (): boolean {
    return !!session.get('admin-header')
  },
  signIn (username: string, password: string) {
    const authValue = Base64.encode(`${username}:${password}Â£`)
    const authHeader = `Basic ${authValue}`
    session.set('admin-header', authHeader)
    events.emit(events.setAuthHeader, authHeader)
  },
  signOut () {
    session.set('admin-header', null)
    events.emit(events.setAuthHeader, null)
  }
}
