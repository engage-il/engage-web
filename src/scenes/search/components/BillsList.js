// @flow
import React, { Component } from 'react'
import { createPaginationContainer, graphql } from 'react-relay'
import type { RelayPaginationProp } from 'react-relay'
import { withRouter } from 'react-router-dom'
import type { ContextRouter } from 'react-router-dom'
import type moment from 'moment'
import { BillCell } from './BillCell'
import { BillAnimation, billRule } from './BillAnimation'
import { LoadMoreButton } from './LoadMoreButton'
import { constants } from '../searchRoute'
import { withLoadMoreArgs, unwrap } from 'shared/relay'
import { session } from 'shared/storage'
import { stylesheet, mixins } from 'shared/styles'
import type { Viewer } from 'shared/types'

function format (date: moment): string {
  return date.format('MMM Do')
}

let BillsList = class BillsList extends Component {
  props: {
    viewer: Viewer,
    animated: Boolean,
    relay: RelayPaginationProp
  } & ContextRouter

  state = {
    disableAnimations: false
  }

  // events
  didClickLoadMore = () => {
    const { relay } = this.props
    if (!relay.hasMore() || relay.isLoading()) {
      return
    }

    relay.loadMore(constants.count, (error: ?Error) => {
      if (error) {
        console.error(`error loading next page: ${error.toString()}`)
      }
    })
  }

  // lifecycle
  componentWillMount () {
    if (this.props.history.action === 'POP') {
      this.setState({ disableAnimations: true })
    }
  }

  componentDidMount () {
    if (this.props.history.action === 'POP') {
      this.setState({ disableAnimations: false })
    }
  }

  componentWillUnmount () {
    const { viewer } = this.props
    if (viewer) {
      session.set('last-search-count', `${viewer.bills.edges.length}`)
    }
  }

  render () {
    const { disableAnimations } = this.state
    const { relay, viewer, animated } = this.props
    const { bills } = viewer
    const { count } = bills
    const { startDate, endDate } = constants

    return <div {...rules.container}>
      <div {...rules.header}>
        <h2>Upcoming Bills</h2>
        <div>{`${format(startDate)} to ${format(endDate)}`}</div>
        <div>{`Found ${count} result${count === 1 ? '' : 's'}.`}</div>
      </div>
      <BillAnimation disable={!animated || disableAnimations}>
        {this.renderBills(unwrap(bills))}
      </BillAnimation>
      <LoadMoreButton
        styles={rules.loadMoreButton}
        hasMore={relay.hasMore()}
        onClick={this.didClickLoadMore}
      />
    </div>
  }

  renderBills (bills): Array<React$Element<*>> {
    return bills.map((bill, i) => {
      return <BillCell key={bill.id} styles={billRule} bill={bill} />
    })
  }
}

BillsList = createPaginationContainer(withRouter(BillsList),
  graphql`
    fragment BillsList_viewer on Viewer {
      bills(
        first: $count, after: $cursor,
        query: $query, from: $startDate, to: $endDate
      ) @connection(key: "BillsList_bills") {
        count
        pageInfo {
          hasNextPage
          endCursor
        }
        edges {
          node {
            id
            ...BillCell_bill
          }
        }
      }
    }
  `,
  withLoadMoreArgs({
    getConnectionFromProps (props) {
      return props.viewer && props.viewer.bills
    },
    query: graphql`
      query BillsListQuery(
        $count: Int!, $cursor: String!,
        $query: String!, $startDate: Time!, $endDate: Time!
      ) {
        viewer {
          ...BillsList_viewer
        }
      }
    `
  })
)

const rules = stylesheet({
  container: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'stretch'
  },
  header: {
    marginBottom: 15,
    '> h2': {
      display: 'inline-block',
      marginBottom: 5,
      ...mixins.mobile({
        marginBottom: 0
      })
    },
    '> div': {
      fontSize: 18,
      ':first-of-type': {
        display: 'inline-block',
        marginLeft: 5
      },
      ...mixins.mobile({
        fontSize: 16,
        ':first-of-type': {
          display: 'none'
        }
      })
    }
  },
  loadMoreButton: {
    alignSelf: 'center',
    marginTop: 30,
    ...mixins.mobile({
      marginTop: 20
    })
  }
})

export { BillsList }
