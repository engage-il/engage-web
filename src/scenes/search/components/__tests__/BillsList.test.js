/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'
import { BillsList } from '../BillsList'
import { routerProps } from 'mocks/routerProps'
import { session } from 'shared/storage'
import { relayPaginationProp } from 'mocks/relayProps'

const { anything } = expect

// mocks
jest.mock('../../searchRoute', () => {
  const moment = require('moment')
  const constants = {
    count: 999,
    startDate: moment().month(4).date(1),
    endDate: moment().month(4).date(7)
  }

  return { constants }
})

// subject
let subject

const defaultProps = {
  viewer: {
    bills: {
      edges: []
    }
  },

  animate: true
}

function loadSubject (props) {
  subject = shallow(<BillsList {...defaultProps} {...props} />).dive().dive()
}

function edges (nodes) {
  return nodes.map((node) => ({ node }))
}

afterEach(() => {
  subject = null
})

describe('#componentWillMount', () => {
  it('disables animations on pop', () => {
    routerProps.history.action = 'POP'
    loadSubject()
    expect(subject).toMatchSnapshot()
  })
})

describe('#componentDidMount', () => {
  it('re-enables animations on pop', () => {
    routerProps.history.action = 'POP'
    loadSubject()
    subject.instance().componentDidMount()
    expect(subject).toMatchSnapshot()
  })
})

describe('#componentWillUnmount', () => {
  it('stores the visible bill count', () => {
    const bills = [{ id: 1 }, { id: 2 }, { id: 3 }]
    loadSubject({viewer: {bills: {edges: edges(bills)}}})
    subject.unmount()
    expect(subject).toMatchSnapshot()
  })
})

describe('#render', () => {
  it('renders correctly', () => {
    loadSubject()
    expect(subject).toMatchSnapshot()
  })

  it('shows the total bill count', () => {
    loadSubject({viewer: {bills: {edges: [], count: 2}}})
    expect(subject).toMatchSnapshot()
  })

  it('shows a bill cell for every bill', () => {
    const bills = [{ id: 1 }, { id: 2 }]
    loadSubject({viewer: {bills: {edges: edges(bills)}}})
    expect(subject).toMatchSnapshot()
  })

  it(`shows the load more button when there's another page`, () => {
    relayPaginationProp.hasMore.mockReturnValue(true)
    loadSubject()
    expect(subject).toMatchSnapshot()
  })
})

describe('on clicking load more', () => {
  it('when there is more to load', () => {
    relayPaginationProp.hasMore.mockReturnValue(true)
    loadSubject()
    subject.find('LoadMoreButton').simulate('click')
    expect(subject).toMatchSnapshot()
  })

  it('when there is no more to load', () => {
    relayPaginationProp.hasMore.mockReturnValue(false)
    loadSubject()
    subject.find('LoadMoreButton').simulate('click')
    expect(subject).toMatchSnapshot()
  })
})

describe('the relay container', () => {
  it('exists', () => {
    expect(BillsList.container).toBeTruthy()
  })
})
