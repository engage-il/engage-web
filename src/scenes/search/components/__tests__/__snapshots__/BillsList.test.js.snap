// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#componentDidMount re-enables animations on pop 1`] = `
<div
  data-css-1szwbq4=""
>
  <div
    data-css-t7v3vx=""
  >
    <h2>
      Upcoming Bills
    </h2>
    <div>
      May 1st to May 7th
    </div>
    <div>
      Found undefined results.
    </div>
  </div>
  <BillAnimation
    disable={true}
  />
  <LoadMoreButton
    onClick={[Function]}
    styles={
      Object {
        "data-css-1n5n5pb": "",
      }
    }
  />
</div>
`;

exports[`#componentWillMount disables animations on pop 1`] = `
<div
  data-css-1szwbq4=""
>
  <div
    data-css-t7v3vx=""
  >
    <h2>
      Upcoming Bills
    </h2>
    <div>
      May 1st to May 7th
    </div>
    <div>
      Found undefined results.
    </div>
  </div>
  <BillAnimation
    disable={true}
  />
  <LoadMoreButton
    onClick={[Function]}
    styles={
      Object {
        "data-css-1n5n5pb": "",
      }
    }
  />
</div>
`;

exports[`#componentWillUnmount stores the visible bill count 1`] = `
<div
  data-css-1szwbq4=""
>
  <div
    data-css-t7v3vx=""
  >
    <h2>
      Upcoming Bills
    </h2>
    <div>
      May 1st to May 7th
    </div>
    <div>
      Found undefined results.
    </div>
  </div>
  <BillAnimation
    disable={true}
  >
    <BillCell
      bill={
        Object {
          "id": 1,
        }
      }
      styles={
        Object {
          "data-css-1a8sov3": "",
        }
      }
    />
    <BillCell
      bill={
        Object {
          "id": 2,
        }
      }
      styles={
        Object {
          "data-css-1a8sov3": "",
        }
      }
    />
    <BillCell
      bill={
        Object {
          "id": 3,
        }
      }
      styles={
        Object {
          "data-css-1a8sov3": "",
        }
      }
    />
  </BillAnimation>
  <LoadMoreButton
    onClick={[Function]}
    styles={
      Object {
        "data-css-1n5n5pb": "",
      }
    }
  />
</div>
`;

exports[`#render renders correctly 1`] = `
<div
  data-css-1szwbq4=""
>
  <div
    data-css-t7v3vx=""
  >
    <h2>
      Upcoming Bills
    </h2>
    <div>
      May 1st to May 7th
    </div>
    <div>
      Found undefined results.
    </div>
  </div>
  <BillAnimation
    disable={true}
  />
  <LoadMoreButton
    onClick={[Function]}
    styles={
      Object {
        "data-css-1n5n5pb": "",
      }
    }
  />
</div>
`;

exports[`#render shows a bill cell for every bill 1`] = `
<div
  data-css-1szwbq4=""
>
  <div
    data-css-t7v3vx=""
  >
    <h2>
      Upcoming Bills
    </h2>
    <div>
      May 1st to May 7th
    </div>
    <div>
      Found undefined results.
    </div>
  </div>
  <BillAnimation
    disable={true}
  >
    <BillCell
      bill={
        Object {
          "id": 1,
        }
      }
      styles={
        Object {
          "data-css-1a8sov3": "",
        }
      }
    />
    <BillCell
      bill={
        Object {
          "id": 2,
        }
      }
      styles={
        Object {
          "data-css-1a8sov3": "",
        }
      }
    />
  </BillAnimation>
  <LoadMoreButton
    onClick={[Function]}
    styles={
      Object {
        "data-css-1n5n5pb": "",
      }
    }
  />
</div>
`;

exports[`#render shows the load more button when there's another page 1`] = `
<div
  data-css-1szwbq4=""
>
  <div
    data-css-t7v3vx=""
  >
    <h2>
      Upcoming Bills
    </h2>
    <div>
      May 1st to May 7th
    </div>
    <div>
      Found undefined results.
    </div>
  </div>
  <BillAnimation
    disable={true}
  />
  <LoadMoreButton
    hasMore={true}
    onClick={[Function]}
    styles={
      Object {
        "data-css-1n5n5pb": "",
      }
    }
  />
</div>
`;

exports[`#render shows the total bill count 1`] = `
<div
  data-css-1szwbq4=""
>
  <div
    data-css-t7v3vx=""
  >
    <h2>
      Upcoming Bills
    </h2>
    <div>
      May 1st to May 7th
    </div>
    <div>
      Found 2 results.
    </div>
  </div>
  <BillAnimation
    disable={true}
  />
  <LoadMoreButton
    onClick={[Function]}
    styles={
      Object {
        "data-css-1n5n5pb": "",
      }
    }
  />
</div>
`;

exports[`on clicking load more when there is more to load 1`] = `
<div
  data-css-1szwbq4=""
>
  <div
    data-css-t7v3vx=""
  >
    <h2>
      Upcoming Bills
    </h2>
    <div>
      May 1st to May 7th
    </div>
    <div>
      Found undefined results.
    </div>
  </div>
  <BillAnimation
    disable={true}
  />
  <LoadMoreButton
    hasMore={true}
    onClick={[Function]}
    styles={
      Object {
        "data-css-1n5n5pb": "",
      }
    }
  />
</div>
`;

exports[`on clicking load more when there is no more to load 1`] = `
<div
  data-css-1szwbq4=""
>
  <div
    data-css-t7v3vx=""
  >
    <h2>
      Upcoming Bills
    </h2>
    <div>
      May 1st to May 7th
    </div>
    <div>
      Found undefined results.
    </div>
  </div>
  <BillAnimation
    disable={true}
  />
  <LoadMoreButton
    hasMore={false}
    onClick={[Function]}
    styles={
      Object {
        "data-css-1n5n5pb": "",
      }
    }
  />
</div>
`;
